svc.go

```diff
@@ -0,0 +1,211 @@
+	opts = append(opts, libp2p.NoListenAddrs)
+	dialer, err := libp2p.New(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	as := &AutoNATService{
+		ctx:    ctx,
+		dialer: dialer,
+		peers:  make(map[peer.ID]struct{}),
+	}
+	h.SetStreamHandler(AutoNATProto, as.handleStream)
+
+	go as.resetPeers()
+
+	return as, nil
+}
+
+func (as *AutoNATService) handleStream(s inet.Stream) {
+	defer s.Close()
+
+	pid := s.Conn().RemotePeer()
+	log.Debugf("New stream from %s", pid.Pretty())
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	var req pb.Message
+	var res pb.Message
+
+	err := r.ReadMsg(&req)
+	if err != nil {
+		log.Debugf("Error reading message from %s: %s", pid.Pretty(), err.Error())
+		s.Reset()
+		return
+	}
+
+	t := req.GetType()
+	if t != pb.Message_DIAL {
+		log.Debugf("Unexpected message from %s: %s (%d)", pid.Pretty(), t.String(), t)
+		s.Reset()
+		return
+	}
+
+	dr := as.handleDial(pid, s.Conn().RemoteMultiaddr(), req.GetDial().GetPeer())
+	res.Type = pb.Message_DIAL_RESPONSE.Enum()
+	res.DialResponse = dr
+
+	err = w.WriteMsg(&res)
+	if err != nil {
+		log.Debugf("Error writing response to %s: %s", pid.Pretty(), err.Error())
+		s.Reset()
+		return
+	}
+}
+
+func (as *AutoNATService) handleDial(p peer.ID, obsaddr ma.Multiaddr, mpi *pb.Message_PeerInfo) *pb.Message_DialResponse {
+	if mpi == nil {
+		return newDialResponseError(pb.Message_E_BAD_REQUEST, "missing peer info")
+	}
+
+	mpid := mpi.GetId()
+	if mpid != nil {
+		mp, err := peer.IDFromBytes(mpid)
+		if err != nil {
+			return newDialResponseError(pb.Message_E_BAD_REQUEST, "bad peer id")
+		}
+
+		if mp != p {
+			return newDialResponseError(pb.Message_E_BAD_REQUEST, "peer id mismatch")
+		}
+	}
+
+	addrs := make([]ma.Multiaddr, 0)
+	seen := make(map[string]struct{})
+
+	// add observed addr to the list of addresses to dial
+	if !as.skipDial(obsaddr) {
+		addrs = append(addrs, obsaddr)
+		seen[obsaddr.String()] = struct{}{}
+	}
+
+	for _, maddr := range mpi.GetAddrs() {
+		addr, err := ma.NewMultiaddrBytes(maddr)
+		if err != nil {
+			log.Debugf("Error parsing multiaddr: %s", err.Error())
+			continue
+		}
+
+		if as.skipDial(addr) {
+			continue
+		}
+
+		str := addr.String()
+		_, ok := seen[str]
+		if ok {
+			continue
+		}
+
+		addrs = append(addrs, addr)
+		seen[str] = struct{}{}
+	}
+
+	if len(addrs) == 0 {
+		return newDialResponseError(pb.Message_E_DIAL_ERROR, "no dialable addresses")
+	}
+
+	return as.doDial(pstore.PeerInfo{ID: p, Addrs: addrs})
+}
+
+func (as *AutoNATService) skipDial(addr ma.Multiaddr) bool {
+	// skip relay addresses
+	_, err := addr.ValueForProtocol(P_CIRCUIT)
+	if err == nil {
+		return true
+	}
+
+	// skip private network (unroutable) addresses
+	if !isPublicAddr(addr) {
+		return true
+	}
+
+	return false
+}
+
+func (as *AutoNATService) doDial(pi pstore.PeerInfo) *pb.Message_DialResponse {
+	// rate limit check
+	as.mx.Lock()
+	_, ok := as.peers[pi.ID]
+	if ok {
```

Comment:
Two comments:

1. Perhaps this n=1 limiting is too optimistic. The first attempt could've failed due to an intermittent issue, not because the address is unroutable. How about allowing n=3 (or configurable) grace attempts per peer?
2. It's confusing that `AutoNATService` and `AmbientAutoNAT` both have a `peers` field, and they're both referred to by `as.peers` in code. I suggest renaming this one to `dialedPeers` to reduce ambiguity.

Reply:
On a related note, both `peers` maps are never cleaned, so they'll keep growing forever. Any thoughts on adding a cleanup routine?

Reply:
> It's confusing that AutoNATService and AmbientAutoNAT both have a peers field, and they're both referred to by as.peers in code.

They serve different purposes, but they are both sequences of peer ids -- hence the name.

Reply:
We _could_ add a comment in the struct though.

Reply:
Implemented configurable throttle in https://github.com/libp2p/go-libp2p-autonat/pull/1/commits/8ea9f1b346487d560c88ce84bc06543a952a75d1

Reply:
> They serve different purposes, but they are both sequences of peer ids -- hence the name.
It took me several reads to understand that `as.peers` referred to different things in the same package. Let's be nice to future maintainers and rename one of them to contextualise it.

Reply:
ok, _fine_. I'll be nice to future self and call the limiter requests.